File: tools/bij/internal/registry/registry.go
package registry

import (
	"fmt"
	"path/filepath"

	"github.com/bijmantra/bij/internal/config"
)

// DefaultRegistry returns the built-in BijMantra development service registry.
func DefaultRegistry(workspace, podman string) []*Service {
	logDir := filepath.Join(workspace, "logs")

	return []*Service{
		{
			Name:           "postgres",
			Label:          "PostgreSQL",
			Type:           ContainerService,
			Group:          ServiceGroupCore,
			Port:           5432,
			ComposeService: "postgres",
			ContainerName:  "bijmantra-postgres",
			Probe: ExecProbe{
				Container: "bijmantra-postgres",
				Cmd:       []string{"pg_isready", "-U", "bijmantra_user", "-d", "bijmantra_db"},
				Runtime:   podman,
			},
			LogSource: PodmanLogSource{Container: "bijmantra-postgres", Runtime: podman},
			State:     StateWaiting,
		},
		{
			Name:            "redis",
			Label:           "Redis",
			Type:            ContainerService,
			Group:           ServiceGroupInfra,
			Port:            6379,
			ComposeService:  "redis",
			ContainerName:   "bijmantra-redis",
			ComposeProfiles: []string{"infra"},
			Probe: ExecProbe{
				Container: "bijmantra-redis",
				Cmd:       []string{"redis-cli", "ping"},
				Runtime:   podman,
			},
			LogSource: PodmanLogSource{Container: "bijmantra-redis", Runtime: podman},
			State:     StateWaiting,
		},
		{
			Name:            "minio",
			Label:           "MinIO",
			Type:            ContainerService,
			Group:           ServiceGroupInfra,
			Port:            9000,
			ComposeService:  "minio",
			ContainerName:   "bijmantra-minio",
			ComposeProfiles: []string{"infra"},
			Probe:           HTTPProbe{URL: "http://localhost:9000/minio/health/live"},
			LogSource:       PodmanLogSource{Container: "bijmantra-minio", Runtime: podman},
			State:           StateWaiting,
		},
		{
			Name:            "meilisearch",
			Label:           "Meilisearch",
			Type:            ContainerService,
			Group:           ServiceGroupInfra,
			Port:            7700,
			ComposeService:  "meilisearch",
			ContainerName:   "bijmantra-meilisearch",
			ComposeProfiles: []string{"infra"},
			Probe:           HTTPProbe{URL: "http://localhost:7700/health"},
			LogSource:       PodmanLogSource{Container: "bijmantra-meilisearch", Runtime: podman},
			State:           StateWaiting,
		},
		{
			Name:            "beingbijmantra",
			Label:           "BeingBijmantra",
			Type:            ContainerService,
			Group:           ServiceGroupAutonomy,
			Experimental:    true,
			Port:            8083,
			ComposeService:  "beingbijmantra-surrealdb",
			ContainerName:   "beingbijmantra-surrealdb",
			ComposeProfiles: []string{"beingbijmantra"},
			Probe:           TCPProbe{Host: "localhost", Port: 8083},
			LogSource:       PodmanLogSource{Container: "beingbijmantra-surrealdb", Runtime: podman},
			State:           StateWaiting,
		},
		{
			Name:      "backend",
			Label:     "Backend API",
			Type:      ProcessService,
			Group:     ServiceGroupCore,
			Port:      8000,
			Cmd:       backendCommand(8000),
			WorkDir:   filepath.Join(workspace, "backend"),
			Probe:     HTTPProbe{URL: "http://localhost:8000/health"},
			LogSource: FileLogSource{Path: filepath.Join(logDir, "backend.log")},
			State:     StateWaiting,
		},
		{
			Name:      "frontend",
			Label:     "Frontend",
			Type:      ProcessService,
			Group:     ServiceGroupCore,
			Port:      5656,
			Cmd:       []string{"bun", "run", "dev"},
			WorkDir:   filepath.Join(workspace, "frontend"),
			Probe:     HTTPProbe{URL: "http://localhost:5656/"},
			LogSource: FileLogSource{Path: filepath.Join(logDir, "frontend.log")},
			State:     StateWaiting,
		},
		{
			Name:            "chloe-gateway",
			Label:           "Chloe Gateway",
			Type:            ContainerService,
			Group:           ServiceGroupAutonomy,
			Experimental:    true,
			Port:            18790,
			ComposeService:  "chloe-gateway",
			ContainerName:   "bijmantra-chloe-gateway",
			ComposeProfiles: []string{"chloe"},
			Probe:           HTTPProbe{URL: "http://127.0.0.1:18790/healthz"},
			LogSource:       PodmanLogSource{Container: "bijmantra-chloe-gateway", Runtime: podman},
			State:           StateWaiting,
		},
		{
			Name:            "chloe-cli",
			Label:           "Chloe CLI",
			Type:            ContainerService,
			Group:           ServiceGroupAutonomy,
			Experimental:    true,
			ComposeService:  "chloe-cli",
			ContainerName:   "bijmantra-chloe-cli",
			ComposeProfiles: []string{"chloe"},
			Probe:           ContainerRunningProbe{Container: "bijmantra-chloe-cli", Runtime: podman},
			LogSource:       PodmanLogSource{Container: "bijmantra-chloe-cli", Runtime: podman},
			State:           StateWaiting,
		},
		{
			Name:            "chloe-sandbox",
			Label:           "Chloe Sandbox",
			Type:            ContainerService,
			Group:           ServiceGroupAutonomy,
			Experimental:    true,
			ComposeService:  "chloe-sandbox",
			ContainerName:   "bijmantra-chloe-sandbox",
			ComposeProfiles: []string{"chloe"},
			Probe:           ContainerRunningProbe{Container: "bijmantra-chloe-sandbox", Runtime: podman},
			LogSource:       PodmanLogSource{Container: "bijmantra-chloe-sandbox", Runtime: podman},
			State:           StateWaiting,
		},
	}
}

// Apply merges config overrides onto the service registry in-place.
func Apply(services []*Service, cfg *config.Config) {
	if cfg == nil {
		return
	}

	for _, svc := range services {
		override, ok := cfg.Services[svc.Name]
		if !ok {
			continue
		}
		hadCommandOverride := override.Cmd != nil

		if override.Port != 0 {
			svc.Port = override.Port
		}
		if hadCommandOverride {
			svc.Cmd = cloneStrings(override.Cmd)
		}
		if override.Port != 0 && !hadCommandOverride && svc.Name == "backend" {
			svc.Cmd = backendCommand(svc.Port)
		}

		refreshProbe(svc)
	}
}

func refreshProbe(svc *Service) {
	switch svc.Name {
	case "minio":
		svc.Probe = HTTPProbe{URL: fmt.Sprintf("http://localhost:%d/minio/health/live", svc.Port)}
	case "meilisearch":
		svc.Probe = HTTPProbe{URL: fmt.Sprintf("http://localhost:%d/health", svc.Port)}
	case "beingbijmantra":
		svc.Probe = TCPProbe{Host: "localhost", Port: svc.Port}
	case "backend":
		svc.Probe = HTTPProbe{URL: fmt.Sprintf("http://localhost:%d/health", svc.Port)}
	case "frontend":
		svc.Probe = HTTPProbe{URL: fmt.Sprintf("http://localhost:%d/", svc.Port)}
	case "chloe-gateway":
		svc.Probe = HTTPProbe{URL: fmt.Sprintf("http://127.0.0.1:%d/healthz", svc.Port)}
	}
}

func backendCommand(port int) []string {
	return []string{
		"uv",
		"run",
		"uvicorn",
		"app.main:app",
		"--reload",
		"--host",
		"0.0.0.0",
		"--port",
		fmt.Sprintf("%d", port),
	}
}

func cloneStrings(values []string) []string {
	if values == nil {
		return nil
	}
	out := make([]string, len(values))
	copy(out, values)
	return out
}

File: crates/bijmantra-runtime/src/registry.rs
use std::path::{Path, PathBuf};
use std::str::FromStr;

use serde::{Deserialize, Serialize};

use crate::config::Config;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ServiceType {
    Container,
    Process,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ServiceGroup {
    Core,
    Infra,
    Autonomy,
}

impl ServiceGroup {
    pub fn includes(self, service: &Service) -> bool {
        service.group <= self
    }
}

impl FromStr for ServiceGroup {
    type Err = anyhow::Error;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "core" => Ok(Self::Core),
            "infra" => Ok(Self::Infra),
            "autonomy" => Ok(Self::Autonomy),
            _ => {
                anyhow::bail!("unknown service group {value:?}; expected core, infra, or autonomy")
            }
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ServiceState {
    Waiting,
    Ready,
    Degraded,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum Probe {
    Tcp {
        host: String,
        port: u16,
    },
    Http {
        url: String,
    },
    Exec {
        runtime: PathBuf,
        container: String,
        cmd: Vec<String>,
    },
    ContainerRunning {
        runtime: PathBuf,
        container: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Service {
    pub name: String,
    pub label: String,
    pub service_type: ServiceType,
    pub group: ServiceGroup,
    pub experimental: bool,
    pub port: Option<u16>,
    pub compose_service: Option<String>,
    pub container_name: Option<String>,
    pub compose_profiles: Vec<String>,
    pub cmd: Vec<String>,
    pub work_dir: Option<PathBuf>,
    pub probe: Option<Probe>,
    pub state: ServiceState,
}

impl Service {
    fn container(
        name: &str,
        label: &str,
        group: ServiceGroup,
        port: Option<u16>,
        compose_service: &str,
        container_name: &str,
        probe: Probe,
    ) -> Self {
        Self {
            name: name.to_string(),
            label: label.to_string(),
            service_type: ServiceType::Container,
            group,
            experimental: false,
            port,
            compose_service: Some(compose_service.to_string()),
            container_name: Some(container_name.to_string()),
            compose_profiles: Vec::new(),
            cmd: Vec::new(),
            work_dir: None,
            probe: Some(probe),
            state: ServiceState::Waiting,
        }
    }

    fn process(
        name: &str,
        label: &str,
        group: ServiceGroup,
        port: u16,
        cmd: Vec<String>,
        work_dir: PathBuf,
        probe: Probe,
    ) -> Self {
        Self {
            name: name.to_string(),
            label: label.to_string(),
            service_type: ServiceType::Process,
            group,
            experimental: false,
            port: Some(port),
            compose_service: None,
            container_name: None,
            compose_profiles: Vec::new(),
            cmd,
            work_dir: Some(work_dir),
            probe: Some(probe),
            state: ServiceState::Waiting,
        }
    }
}

pub fn default_registry(workspace: &Path, podman: &Path) -> Vec<Service> {
    let mut services = vec![
        Service::container(
            "postgres",
            "PostgreSQL",
            ServiceGroup::Core,
            Some(5432),
            "postgres",
            "bijmantra-postgres",
            Probe::Exec {
                runtime: podman.to_path_buf(),
                container: "bijmantra-postgres".to_string(),
                cmd: vec![
                    "pg_isready".to_string(),
                    "-U".to_string(),
                    "bijmantra_user".to_string(),
                    "-d".to_string(),
                    "bijmantra_db".to_string(),
                ],
            },
        ),
        Service::container(
            "redis",
            "Redis",
            ServiceGroup::Infra,
            Some(6379),
            "redis",
            "bijmantra-redis",
            Probe::Exec {
                runtime: podman.to_path_buf(),
                container: "bijmantra-redis".to_string(),
                cmd: vec!["redis-cli".to_string(), "ping".to_string()],
            },
        ),
        Service::container(
            "minio",
            "MinIO",
            ServiceGroup::Infra,
            Some(9000),
            "minio",
            "bijmantra-minio",
            Probe::Http {
                url: "http://localhost:9000/minio/health/live".to_string(),
            },
        ),
        Service::container(
            "meilisearch",
            "Meilisearch",
            ServiceGroup::Infra,
            Some(7700),
            "meilisearch",
            "bijmantra-meilisearch",
            Probe::Http {
                url: "http://localhost:7700/health".to_string(),
            },
        ),
        Service::container(
            "beingbijmantra",
            "BeingBijmantra",
            ServiceGroup::Autonomy,
            Some(8083),
            "beingbijmantra-surrealdb",
            "beingbijmantra-surrealdb",
            Probe::Tcp {
                host: "localhost".to_string(),
                port: 8083,
            },
        ),
        Service::process(
            "backend",
            "Backend API",
            ServiceGroup::Core,
            8000,
            backend_command(8000),
            workspace.join("backend"),
            Probe::Http {
                url: "http://localhost:8000/health".to_string(),
            },
        ),
        Service::process(
            "frontend",
            "Frontend",
            ServiceGroup::Core,
            5656,
            vec!["bun".to_string(), "run".to_string(), "dev".to_string()],
            workspace.join("frontend"),
            Probe::Http {
                url: "http://localhost:5656/".to_string(),
            },
        ),
        Service::container(
            "chloe-gateway",
            "Chloe Gateway",
            ServiceGroup::Autonomy,
            Some(18790),
            "chloe-gateway",
            "bijmantra-chloe-gateway",
            Probe::Http {
                url: "http://127.0.0.1:18790/healthz".to_string(),
            },
        ),
        Service::container(
            "chloe-cli",
            "Chloe CLI",
            ServiceGroup::Autonomy,
            None,
            "chloe-cli",
            "bijmantra-chloe-cli",
            Probe::ContainerRunning {
                runtime: podman.to_path_buf(),
                container: "bijmantra-chloe-cli".to_string(),
            },
        ),
        Service::container(
            "chloe-sandbox",
            "Chloe Sandbox",
            ServiceGroup::Autonomy,
            None,
            "chloe-sandbox",
            "bijmantra-chloe-sandbox",
            Probe::ContainerRunning {
                runtime: podman.to_path_buf(),
                container: "bijmantra-chloe-sandbox".to_string(),
            },
        ),
    ];

    for service in &mut services {
        match service.name.as_str() {
            "redis" | "minio" | "meilisearch" => {
                service.compose_profiles = vec!["infra".to_string()]
            }
            "beingbijmantra" => {
                service.experimental = true;
                service.compose_profiles = vec!["beingbijmantra".to_string()];
            }
            "chloe-gateway" | "chloe-cli" | "chloe-sandbox" => {
                service.experimental = true;
                service.compose_profiles = vec!["chloe".to_string()];
            }
            _ => {}
        }
    }

    services
}

pub fn apply_config(services: &mut [Service], config: &Config) {
    for service in services {
        let Some(override_config) = config.services.get(&service.name) else {
            continue;
        };

        let had_cmd_override = override_config.cmd.is_some();
        if let Some(port) = override_config.port {
            service.port = Some(port);
        }
        if let Some(cmd) = &override_config.cmd {
            service.cmd = cmd.clone();
        }
        if service.name == "backend" && service.port.is_some() && !had_cmd_override {
            service.cmd = backend_command(service.port.unwrap_or(8000));
        }
        refresh_probe(service);
    }
}

pub fn services_for_group(services: &[Service], group: ServiceGroup) -> Vec<Service> {
    services
        .iter()
        .filter(|service| group.includes(service))
        .cloned()
        .collect()
}

fn refresh_probe(service: &mut Service) {
    let Some(port) = service.port else {
        return;
    };

    service.probe = match service.name.as_str() {
        "minio" => Some(Probe::Http {
            url: format!("http://localhost:{port}/minio/health/live"),
        }),
        "meilisearch" => Some(Probe::Http {
            url: format!("http://localhost:{port}/health"),
        }),
        "beingbijmantra" => Some(Probe::Tcp {
            host: "localhost".to_string(),
            port,
        }),
        "backend" => Some(Probe::Http {
            url: format!("http://localhost:{port}/health"),
        }),
        "frontend" => Some(Probe::Http {
            url: format!("http://localhost:{port}/"),
        }),
        "chloe-gateway" => Some(Probe::Http {
            url: format!("http://127.0.0.1:{port}/healthz"),
        }),
        _ => service.probe.clone(),
    };
}

fn backend_command(port: u16) -> Vec<String> {
    vec![
        "uv".to_string(),
        "run".to_string(),
        "uvicorn".to_string(),
        "app.main:app".to_string(),
        "--reload".to_string(),
        "--host".to_string(),
        "0.0.0.0".to_string(),
        "--port".to_string(),
        port.to_string(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn infra_group_includes_core_and_infra_services() {
        let services = default_registry(Path::new("."), Path::new("podman"));
        let selected = services_for_group(&services, ServiceGroup::Infra);
        let names: Vec<_> = selected
            .iter()
            .map(|service| service.name.as_str())
            .collect();

        assert!(names.contains(&"postgres"));
        assert!(names.contains(&"backend"));
        assert!(names.contains(&"frontend"));
        assert!(names.contains(&"redis"));
        assert!(!names.contains(&"chloe-gateway"));
    }

    #[test]
    fn backend_port_override_refreshes_command_and_probe() {
        let mut services = default_registry(Path::new("."), Path::new("podman"));
        let mut config = Config::default();
        config.services.insert(
            "backend".to_string(),
            crate::config::ServiceOverride {
                port: Some(18000),
                cmd: None,
            },
        );

        apply_config(&mut services, &config);

        let backend = services
            .iter()
            .find(|service| service.name == "backend")
            .unwrap();
        assert!(backend.cmd.contains(&"18000".to_string()));
        assert_eq!(
            backend.probe,
            Some(Probe::Http {
                url: "http://localhost:18000/health".to_string()
            })
        );
    }
}
